service/worker/deployment/deployment_workflow.go

```diff
@@ -0,0 +1,150 @@
+	DeploymentTaskQueue struct {
+		FirstPollerTimeStamp *timestamppb.Timestamp
+		TaskQueue            TaskQueue
+	}
+
+	// DeploymentWorkflowArgs represents the arguments passed for a DeploymentWorkflow
+	DeploymentWorkflowArgs struct {
+		NamespaceName string
+		NamespaceID   string
+		TaskQueues    []*DeploymentTaskQueue // required for CAN
+	}
+
+	DeploymentLocalState struct {
+		NamespaceName  string
+		NamespaceID    string
+		DeploymentName string
+		BuildID        string
+		TaskQueues     []*DeploymentTaskQueue // All the task queues associated with this buildID/deployment
+	}
+
+	// DeploymentWorkflowRunner holds the local state for a deployment workflow
+	DeploymentWorkflowRunner struct {
+		ctx                  workflow.Context
+		a                    *DeploymentActivities
+		logger               sdklog.Logger
+		metrics              sdkclient.MetricsHandler
+		deploymentlocalState *DeploymentLocalState
+	}
+)
+
+var (
+	defaultActivityOptions = workflow.ActivityOptions{
+		ScheduleToCloseTimeout: 1 * time.Hour,
+		StartToCloseTimeout:    30 * time.Second,
+		RetryPolicy: &temporal.RetryPolicy{
+			InitialInterval: 1 * time.Second,
+			MaximumInterval: 60 * time.Second,
+		},
+	}
+)
+
+const (
+	UpdateDeploymentSignalName        = "update_deployment"
+	UpdateDeploymentBuildIDSignalName = "update_deployment_build_id"
+
+	DeploymentWorkflowIDPrefix = "temporal-sys-deployment:"
+)
+
+func DeploymentWorkflow(ctx workflow.Context, deploymentWorkflowArgs DeploymentWorkflowArgs) error {
+	deploymentWorkflowRunner := &DeploymentWorkflowRunner{
+		deploymentlocalState: &DeploymentLocalState{
+			NamespaceName:  deploymentWorkflowArgs.NamespaceName,
+			NamespaceID:    deploymentWorkflowArgs.NamespaceID,
+			DeploymentName: "", // TODO Shivam - extract the Deployment
+			BuildID:        "", // TODO Shivam - extract BuildID from the workflowID
+			TaskQueues:     deploymentWorkflowArgs.TaskQueues,
+		},
+		ctx:     ctx,
+		logger:  sdklog.With(workflow.GetLogger(ctx), "wf-namespace", deploymentWorkflowArgs.NamespaceName),
+		metrics: workflow.GetMetricsHandler(ctx).WithTags(map[string]string{"namespace": deploymentWorkflowArgs.NamespaceName}),
+	}
+	return deploymentWorkflowRunner.run()
+}
+
+func (d *DeploymentWorkflowRunner) run() error {
+
+	// Set up Query Handlers here:
+
+	// Fetch signal channels
+	updateDeploymentSignalChannel := workflow.GetSignalChannel(d.ctx, UpdateDeploymentSignalName)
+	updateBuildIDSignalChannel := workflow.GetSignalChannel(d.ctx, UpdateDeploymentBuildIDSignalName)
+
+	selector := workflow.NewSelector(d.ctx)
+	selector.AddReceive(updateDeploymentSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		// Process Signal
+
+	})
+	selector.AddReceive(updateBuildIDSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		// Process Signal
+	})
+
+	// async draining before CAN
+	for !workflow.GetInfo(d.ctx).GetContinueAsNewSuggested() || selector.HasPending() {
```

Comment:
One thing that we did with schedules that we should probably do here: add a signal to force the wf to continue as new immediately (well, if no other signals are pending). That can help with upgrades/downgrades.

Reply:
I am in agreement with this. This will also help in writing simpler unit test imo. I shall make the changes.